lib/colocated-broccoli-plugin.js

```diff
@@ -132,29 +167,36 @@ module.exports = class ColocatedTemplateProcessor extends Plugin {
 
       jsContents = prefix + jsContents;
 
-      let outputPath = path.join(this.outputPath, backingClassPath);
-
-      // TODO: don't speculatively mkdirSync (likely do in a try/catch with ENOENT)
-      mkdirp.sync(path.dirname(outputPath));
-      fs.writeFileSync(outputPath, jsContents, { encoding: 'utf8' });
-    });
-
-    filesToCopy.forEach(filePath => {
-      let inputPath = path.join(this.inputPaths[0], filePath);
-      let outputPath = path.join(this.outputPath, filePath);
-
-      // avoid copying file over top of a previously written one
-      if (fs.existsSync(outputPath)) {
-        return;
+      let jsOutputPath = path.join(this.outputPath, backingClassPath);
+
+      switch (method) {
+        case 'unlink': {
+          if (filePathParts.ext === '.hbs' && hasBackingClass) {
+            fs.writeFileSync(jsOutputPath, originalJsContents, { encoding: 'utf8' });
+          } else if (filePathParts.ext !== '.hbs' && hasTemplate) {
+            fs.writeFileSync(jsOutputPath, jsContents, { encoding: 'utf8' });
+          } else {
+            // Copied from https://github.com/stefanpenner/fs-tree-diff/blob/v2.0.1/lib/index.ts#L38-L68
+            try {
+              fs.unlinkSync(jsOutputPath);
+            } catch (e) {
+              if (typeof e === 'object' && e !== null && e.code === 'ENOENT') {
+                return;
+              }
+              throw e;
+            }
+          }
+          break;
+        }
+        case 'change':
+        case 'create': {
+          fs.writeFileSync(jsOutputPath, jsContents, { encoding: 'utf8' });
+          break;
+        }
+        default: {
+          throw new Error(`Unexpected operation ${JSON.stringify([method, relativePath])}`);
```

Comment:
Tweak this to make it a bit more actionable, we don't actually expect this to be hit by any user but if it does they should know that they should file an issue against the repo here...